File: module_utils/agent_identity.py
"""Agent identity helper — Track B (2026-04-26).

A tiny, dependency-light client_credentials helper used by callers that need
to talk to Bone with a Bearer JWT instead of the retired X-API-Key channel.

Used by:
    - library/nos_state.py (state-manager role end-of-run reporter)
    - any future Ansible module that POSTs to Bone

Tokens are cached on disk under ~/.nos/agent-tokens/<client_id>.json so
multiple ansible-playbook runs in the same TTL window don't hammer Authentik
for fresh tokens. The cache is rejected if the token is within
``REFRESH_BEFORE_EXPIRY_SECONDS`` of expiry — clock skew across hosts is
small and ansible runs rarely exceed a few minutes anyway.

This module deliberately avoids:
  - PyJWT dependency on the caller side (we don't VERIFY tokens here, we
    only fetch them — Authentik is the trust anchor, Bone is the verifier)
  - async/await (Ansible's run_module entrypoints are sync; making them
    async forces an event loop in every call site)

The only third-party import is `urllib`/`json`, which stdlib ships.
"""

from __future__ import annotations

import json
import os
import threading
import time
import urllib.error
import urllib.parse
import urllib.request

# When the cached token has fewer than this many seconds left we refresh.
# Authentik's default access token validity is 12h (Track B decision O3),
# so 60s leaves plenty of buffer for slow networks.
REFRESH_BEFORE_EXPIRY_SECONDS = 60

# Token cache directory. Created lazily; honors $NOS_AGENT_TOKEN_DIR override
# so test suites can sandbox.
TOKEN_DIR = os.environ.get(
    "NOS_AGENT_TOKEN_DIR",
    os.path.expanduser("~/.nos/agent-tokens"),
)

_LOCK = threading.Lock()


class AgentIdentityError(RuntimeError):
    """Raised when token acquisition fails for any reason."""


def _ensure_token_dir() -> None:
    if not os.path.isdir(TOKEN_DIR):
        os.makedirs(TOKEN_DIR, mode=0o700, exist_ok=True)


def _cache_path(client_id: str) -> str:
    safe = "".join(c for c in client_id if c.isalnum() or c in "-_") or "default"
    return os.path.join(TOKEN_DIR, f"{safe}.json")


def _load_cached(client_id: str) -> dict | None:
    path = _cache_path(client_id)
    try:
        with open(path, "r", encoding="utf-8") as fh:
            data = json.load(fh)
    except (OSError, ValueError):
        return None
    expires_at = data.get("expires_at", 0)
    if expires_at - time.time() < REFRESH_BEFORE_EXPIRY_SECONDS:
        return None
    return data


def _store_cache(client_id: str, payload: dict) -> None:
    _ensure_token_dir()
    tmp = _cache_path(client_id) + ".tmp"
    with open(tmp, "w", encoding="utf-8") as fh:
        json.dump(payload, fh)
    os.chmod(tmp, 0o600)
    os.replace(tmp, _cache_path(client_id))


def _fetch_token(token_url: str, client_id: str, client_secret: str,
                 scopes: list[str]) -> dict:
    """POST to Authentik's /token endpoint with client_credentials grant."""
    body = urllib.parse.urlencode({
        "grant_type": "client_credentials",
        "client_id": client_id,
        "client_secret": client_secret,
        "scope": " ".join(scopes),
    }).encode("utf-8")
    req = urllib.request.Request(
        token_url,
        data=body,
        headers={
            "Content-Type": "application/x-www-form-urlencoded",
            "Accept": "application/json",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            data = json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as exc:
        body_txt = exc.read().decode("utf-8", errors="replace")[:500]
        raise AgentIdentityError(
            f"Authentik {exc.code} on {token_url}: {body_txt}"
        ) from exc
    except urllib.error.URLError as exc:
        raise AgentIdentityError(f"Authentik unreachable: {exc.reason}") from exc

    access_token = data.get("access_token")
    if not access_token:
        raise AgentIdentityError(f"Authentik returned no access_token: {data}")
    expires_in = int(data.get("expires_in", 3600))
    return {
        "access_token": access_token,
        "token_type": data.get("token_type", "Bearer"),
        "scope": data.get("scope", " ".join(scopes)),
        "expires_at": int(time.time()) + expires_in,
    }


def get_token(token_url: str, client_id: str, client_secret: str,
              scopes: list[str]) -> str:
    """Return a valid access token for the given OIDC client.

    On cache miss / expiry, fetches a new token from Authentik. Thread-safe
    within a single Python process; cross-process safety is achieved by the
    O_TRUNC + os.replace dance in _store_cache.
    """
    if not token_url or not client_id or not client_secret:
        raise AgentIdentityError(
            "agent_identity.get_token requires token_url, client_id, client_secret"
        )
    with _LOCK:
        cached = _load_cached(client_id)
        if cached:
            return cached["access_token"]
        payload = _fetch_token(token_url, client_id, client_secret, scopes)
        _store_cache(client_id, payload)
        return payload["access_token"]


def authorization_header(token_url: str, client_id: str, client_secret: str,
                         scopes: list[str]) -> dict[str, str]:
    """Convenience: returns ``{"Authorization": "Bearer <jwt>"}``.

    Most callers want the header dict, not the raw token. Pre-formatted so
    they can update the requests/httpx headers in one line.
    """
    token = get_token(token_url, client_id, client_secret, scopes)
    return {"Authorization": f"Bearer {token}"}


def derive_token_url(authentik_domain: str) -> str:
    """Return the Authentik client_credentials token endpoint for a domain.

    Authentik exposes this URL at the global OAuth2 root (not per-application),
    matching the issuer mode we configure in 30-agent-clients.yaml.j2.
    """
    return f"https://{authentik_domain.rstrip('/')}/application/o/token/"


def invalidate(client_id: str) -> bool:
    """Delete the cached token for this client. Returns True if removed."""
    path = _cache_path(client_id)
    try:
        os.unlink(path)
        return True
    except FileNotFoundError:
        return False

File: files/anatomy/wing/app/Core/AgentIdentity.php
<?php
/**
 * Wing -> Bone OIDC client_credentials helper.
 *
 * Mirror of module_utils/agent_identity.py - same on-disk cache layout
 * (~/.nos/agent-tokens/<client_id>.json), same Authentik token endpoint,
 * same refresh window. PHP-side Wing uses this when its read-only models
 * need to GET /api/state/services from Bone.
 *
 * No Composer deps - cURL + json_encode are stdlib so the same code works
 * on macOS Homebrew PHP and Alpine FPM containers.
 *
 * Track B (2026-04-26).
 */

declare(strict_types=1);

namespace App\Core;

use RuntimeException;

final class AgentIdentity
{
    private const REFRESH_BEFORE_EXPIRY_SECONDS = 60;

    private string $tokenUrl;
    private string $clientId;
    private string $clientSecret;
    /** @var list<string> */
    private array $scopes;
    private string $cacheDir;

    /**
     * @param list<string> $scopes
     */
    public function __construct(
        string $tokenUrl,
        string $clientId,
        string $clientSecret,
        array $scopes,
        ?string $cacheDir = null
    ) {
        $this->tokenUrl = $tokenUrl;
        $this->clientId = $clientId;
        $this->clientSecret = $clientSecret;
        $this->scopes = $scopes;
        $this->cacheDir = $cacheDir
            ?? ($_ENV['NOS_AGENT_TOKEN_DIR']
                ?? (getenv('HOME') . '/.nos/agent-tokens'));
    }

    /**
     * Build from environment / .env vars written by pazny.wing.
     * Required env keys: AUTHENTIK_DOMAIN, WING_AGENT_CLIENT_ID,
     * WING_AGENT_CLIENT_SECRET, WING_AGENT_SCOPES.
     */
    public static function fromEnv(): self
    {
        $domain = getenv('AUTHENTIK_DOMAIN') ?: 'auth.dev.local';
        $clientId = getenv('WING_AGENT_CLIENT_ID') ?: 'nos-wing';
        $clientSecret = getenv('WING_AGENT_CLIENT_SECRET') ?: '';
        $scopesRaw = getenv('WING_AGENT_SCOPES') ?: 'nos:state:read';
        $scopes = preg_split('/\s+/', trim($scopesRaw)) ?: [];
        $tokenUrl = "https://{$domain}/application/o/token/";
        return new self($tokenUrl, $clientId, $clientSecret, $scopes);
    }

    public function getToken(): string
    {
        $cached = $this->loadCached();
        if ($cached !== null) {
            return $cached['access_token'];
        }
        $fresh = $this->fetchToken();
        $this->storeCache($fresh);
        return $fresh['access_token'];
    }

    /** @return array{Authorization: string} */
    public function authorizationHeader(): array
    {
        return ['Authorization' => 'Bearer ' . $this->getToken()];
    }

    public function invalidate(): bool
    {
        $path = $this->cachePath();
        if (!is_file($path)) {
            return false;
        }
        return @unlink($path);
    }

    // -- internals -----------------------------------------------------------

    private function cachePath(): string
    {
        $safe = preg_replace('/[^A-Za-z0-9_-]/', '', $this->clientId) ?: 'default';
        return $this->cacheDir . '/' . $safe . '.json';
    }

    /** @return array{access_token: string, expires_at: int}|null */
    private function loadCached(): ?array
    {
        $path = $this->cachePath();
        if (!is_file($path)) {
            return null;
        }
        $raw = @file_get_contents($path);
        if ($raw === false) {
            return null;
        }
        $data = json_decode($raw, true);
        if (!is_array($data) || !isset($data['access_token'], $data['expires_at'])) {
            return null;
        }
        if ((int)$data['expires_at'] - time() < self::REFRESH_BEFORE_EXPIRY_SECONDS) {
            return null;
        }
        return $data;
    }

    /**
     * @param array{access_token: string, expires_at: int} $payload
     */
    private function storeCache(array $payload): void
    {
        if (!is_dir($this->cacheDir)) {
            @mkdir($this->cacheDir, 0700, true);
        }
        $tmp = $this->cachePath() . '.tmp';
        file_put_contents($tmp, json_encode($payload, JSON_UNESCAPED_SLASHES));
        @chmod($tmp, 0600);
        @rename($tmp, $this->cachePath());
    }

    /** @return array{access_token: string, expires_at: int} */
    private function fetchToken(): array
    {
        if ($this->clientSecret === '') {
            throw new RuntimeException(
                'AgentIdentity: WING_AGENT_CLIENT_SECRET is empty. '
                . 'Run the playbook so the wing role writes ~/wing/.env.'
            );
        }
        $body = http_build_query([
            'grant_type'    => 'client_credentials',
            'client_id'     => $this->clientId,
            'client_secret' => $this->clientSecret,
            'scope'         => implode(' ', $this->scopes),
        ]);

        $ch = curl_init($this->tokenUrl);
        curl_setopt_array($ch, [
            CURLOPT_POST           => true,
            CURLOPT_POSTFIELDS     => $body,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => 10,
            CURLOPT_HTTPHEADER     => [
                'Content-Type: application/x-www-form-urlencoded',
                'Accept: application/json',
            ],
        ]);
        $resp = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_RESPONSE_CODE);
        $err = curl_error($ch);
        curl_close($ch);

        if ($resp === false) {
            throw new RuntimeException("Authentik unreachable: {$err}");
        }
        if ($httpCode >= 400) {
            $snippet = substr((string)$resp, 0, 500);
            throw new RuntimeException("Authentik {$httpCode}: {$snippet}");
        }
        $data = json_decode((string)$resp, true);
        if (!is_array($data) || empty($data['access_token'])) {
            throw new RuntimeException('Authentik returned no access_token');
        }
        $expiresIn = (int)($data['expires_in'] ?? 3600);
        return [
            'access_token' => (string)$data['access_token'],
            'expires_at'   => time() + $expiresIn,
        ];
    }
}
